crates/iceberg/src/spec/view_metadata_builder.rs

```diff
@@ -0,0 +1,1566 @@
+            view_version
+        };
+
+        if !self
+            .metadata
+            .schemas
+            .contains_key(&view_version.schema_id())
+        {
+            return Err(Error::new(
+                ErrorKind::DataInvalid,
+                format!(
+                    "Cannot add version with unknown schema: {}",
+                    view_version.schema_id()
+                ),
+            ));
+        }
+
+        require_unique_dialects(&view_version)?;
+
+        // ToDo Discuss: This check is not present in Java.
+        // The `TableMetadataBuilder` uses these checks in multiple places - also in Java.
+        // If we think delayed requests are a problem, I think we should also add it here.
+        if let Some(last) = self.metadata.version_log.last() {
+            // commits can happen concurrently from different machines.
+            // A tolerance helps us avoid failure for small clock skew
+            if view_version.timestamp_ms() - last.timestamp_ms() < -ONE_MINUTE_MS {
+                return Err(Error::new(
+                    ErrorKind::DataInvalid,
+                    format!(
+                        "Invalid snapshot timestamp {}: before last snapshot timestamp {}",
+                        view_version.timestamp_ms(),
+                        last.timestamp_ms()
+                    ),
+                ));
+            }
+        }
+
+        self.metadata
+            .versions
+            .insert(version_id, Arc::new(view_version.clone()));
+
+        let view_version = if let Some(last_added_schema_id) = self.last_added_schema_id {
+            if view_version.schema_id() == last_added_schema_id {
+                view_version.with_schema_id(Self::LAST_ADDED)
+            } else {
+                view_version
+            }
+        } else {
+            view_version
+        };
+        self.changes
+            .push(ViewUpdate::AddViewVersion { view_version });
+
+        self.last_added_version_id = Some(version_id);
+
+        Ok(version_id)
+    }
+
+    fn reuse_or_create_new_view_version_id(&self, new_view_version: &ViewVersion) -> i32 {
+        self.metadata
+            .versions
+            .iter()
+            .find_map(|(id, other_version)| {
+                new_view_version
+                    .behaves_identical_to(other_version)
+                    .then_some(*id)
+            })
+            .unwrap_or_else(|| {
+                self.get_highest_view_version_id()
+                    .map(|id| id + 1)
+                    // ToDo Discuss: In Java this uses `new_view_version.version_id()` instead.
+                    // I believe 0 is more appropriate here, as the first version should always be 0.
+                    // The TableMetadataBuilder also uses 0 for partition specs (and other `reuse_or_create_*` functions).
+                    // Consistent behaviour between the two builders is desirable.
+                    .unwrap_or(INITIAL_VIEW_VERSION_ID)
```

Comment:
CC @Fokko @nastra

Reply:
I don't think there's a requirement from the spec that would dictate whether version id needs to start at 0 or 1 but we actually have code in Java that interpret 1 as the initial view version id: https://github.com/apache/iceberg/blob/6f1517546aceee19e10f38e36df7b072267fe6db/api/src/main/java/org/apache/iceberg/view/ViewVersion.java#L68

Reply:
Oh, so we should default to 1 then too.
If Java depends on this, we might want to add it to the spec?